### ISSUE TITLE
Multiple depots

### ISSUE THREAD

**Issue by N-Wouda** [PyVRP/PyVRP#305]
#### Is your feature request related to a problem? Please describe

N/A

#### Describe the solution you'd like

Support for multiple depots. This is discussed already a little bit in #5, but there is no issue dedicated to just this extension.

#### Describe alternatives you have considered

N/A

#### Additional context

N/A

**Comment by N-Wouda** [comment]
"The" multi-depot VRP has vehicles that belong to different depots. Each vehicle starts and ends at a fixed depot, but that depot need not be the same for different vehicles. I propose we start with this classic variant.

Other variants also allow travel between depots, but we need not support that out of the box because then it gets tricky with e.g. empty routes.

**Comment by N-Wouda** [comment]
An earlier PR #274 removed several `0` in favour of a depot field on the vehicle types. That's a start, but the work is still unfinished, mostly because `crossover::greedyRepair` is not currently able to handle different vehicle types/depots. I suggest we fix that by splitting off `greedyRepair` in the manner of #182. We might want to add that to a `pyvrp.repair` module.

This issue becomes a lot easier if we can figure out #182 first.

**Comment by N-Wouda** [comment]
Not sure what to do about BPD.

**Comment by N-Wouda** [comment]
#360 gets us like 90% of the way there. All we still have to figure out is:
- What to do about BPD. It has no way of knowing how many depots there are, because we removed its access to the data instance a while back. There are IMO two solutions:
  - Count the depots. Since these are always unassigned, that means we will never have 100% broken pairs again, but I don't think that's a problem at all.
  - Have BPD take the data instance as an argument. That feels more heavy-handed to me.
- Find benchmark instances and figure out how to read them with VRPLIB. Should not be too hard; I know there are some around. But they'll probably not be in VRPLIB format.

**Comment by leonlan** [comment]
Here are benchmark instances for:
- [MDVRPTW](http://www.vrp-rep.org/variants/item/mdvrptw.html): Vidal uses 28 instances from 360-960 customers. Cordeau proposes 20 instances with 48-288 customers (probably not useful). See [Vidal et al. (2012)](https://pubsonline.informs.org/doi/abs/10.1287/opre.1120.1048?casa_token=Y3sTFfPqZF0AAAAA:Q639iEOqjCcAT2V1j6XYrGjtt8PcGGZcnSHZjHAgRA6M7FWOODeXNtn3dUqjnYHHyw0BG9ZjO-1PDw) for more details about MDVRP.

Extending with VRPLIB can be done fairly easily:
- Extend the `DEPOT_SECTION` to include more depots.
- Add `VEHICLE_DEPOT_SECTION`, or more generally, `VEHICLE_START_DEPOT_SECTION` and `VEHICLE_END_DEPOT_SECTION` sections.

**Comment by N-Wouda** [comment]
> * [MDVRPTW](http://www.vrp-rep.org/variants/item/mdvrptw.html): Vidal uses 28 instances from 360-960 customers. Cordeau proposes 20 instances with 48-288 customers (probably not useful). See [Vidal et al. (2012)](https://pubsonline.informs.org/doi/abs/10.1287/opre.1120.1048?casa_token=Y3sTFfPqZF0AAAAA:Q639iEOqjCcAT2V1j6XYrGjtt8PcGGZcnSHZjHAgRA6M7FWOODeXNtn3dUqjnYHHyw0BG9ZjO-1PDw) for more details about MDVRP.

I just looked at VRP-REP for instances, and also settled on the 28 instances of Vidal (2013). Those are sufficiently large that it's an interesting comparison.

> * Add `VEHICLE_DEPOT_SECTION`, or more generally, `VEHICLE_START_DEPOT_SECTION` and `VEHICLE_END_DEPOT_SECTION` sections.

I think we should do the former, because there isn't anything planned for travel between depots (and we can revisit alternatives once we might want to do that).

**Comment by N-Wouda** [comment]
I'm working on relevant benchmark instances in https://github.com/PyVRP/Instances/pull/6.

**Comment by N-Wouda** [comment]
> What to do about BPD. It has no way of knowing how many depots there are, because we removed its access to the data instance a while back. There are IMO two solutions:
> * Count the depots. Since these are always unassigned, that means we will never have 100% broken pairs again, but I don't think that's a problem at all.
> * Have BPD take the data instance as an argument. That feels more heavy-handed to me, since it would also require changes to (sub-)population, as that currently does not have the data instance available either. I'd rather we avoid this.

In #411 I went with the first solution.

**Comment by N-Wouda** [comment]
So, benchmarks. We have our instances now. Vidal et al. (2013) used the following stopping criterion:
>To save computation time, the termination criterion was reduced to 2500 iterations without improvement, and a time limit of 5h is imposed.

That seems rather long. Do we also want to do that? For a fair comparison we likely have to, but a fair comparison is already problematic because we do not know the paper's CPU details other than an "Intel Xeon 2.93 GHz processor" (and thus have no clear way forward on how to scale the time limit).

Instead, I suggest we adopt a fixed runtime of 2h, on a CPU with 2000 PassMark score. That seems like a good standard setup, and matches most of our other benchmarking. Unambiguous, and unproblematic after a few releases for backwards comparisons.

@leonlan @wouterkool any thoughts on this?

**Comment by leonlan** [comment]
Yeah, I think our benchmarks (except CVRP) should stick to the 2h, 2000 PassMark, 10 seeds convention. 

Another point is the precision/rounding convention of benchmarks. We should stick to _one_ rounding convention (except for CVRP), which I think should be DIMACS (since it's already used for VRPTW/PCVRPTW). This also means we assume "integer" data and no double precision.

**Comment by N-Wouda** [comment]
> Another point is the precision/rounding convention of benchmarks. We should stick to _one_ rounding convention (except for CVRP), which I think should be DIMACS (since it's already used for VRPTW/PCVRPTW). This also means we assume "integer" data and no double precision.

Yeah, so, about that. MDVRPTW is full double precision, because I couldn't get the scale right without it. It probably overflows somewhere during the run if I scale everything up to integers. Or we accept that there's precision loss, use DIMACS, and then no longer facilitate a comparison to the literature. That's also possible.

**Comment by leonlan** [comment]
I think we should drop the comparison to the literature. We can still compare to the literature to get a sense of how we're doing (like you did now), but it's easiest if we standardize the benchmarks in our own desired way.

**Comment by N-Wouda** [comment]
Alright, we can now write MDVRPTW solutions (or instances with non-homogeneous vehicles in general), and support any mix of the max runtime, max iterations, and no improvement stopping criteria. I think we're good to go for a little benchmark!

**Comment by N-Wouda** [comment]
### Benchmark on MDVRPTW

Current main (b7a5efc3ec50d611343f99286c4bdd665856883e; benchmark settings; job ID 6982597):
```

```

**Comment by N-Wouda** [comment]
I checked, and I don't think much has been done on MDVRPTW since the reference paper from 2013. I found a few papers, but none that directly benchmarked on our benchmark instances. So I think comparing against Vidal et al. (2013) as "current BKS" is reasonable. If we later find papers that have already improved those BKS we can always update the comparison.
